File: src/views/books.rs
use crate::models::_entities::books::Model;
use crate::models::books::BookTree;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct BookResponse {
    pub id: i32,
    pub title: String,
    pub description: Option<String>,
    pub cover_image: Option<String>,
    pub parent_id: Option<i32>,
    pub sort_order: Option<i32>,
    pub is_public: Option<bool>,
    pub user_id: i32,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

impl From<Model> for BookResponse {
    fn from(book: Model) -> Self {
        Self {
            id: book.id,
            title: book.title,
            description: book.description,
            cover_image: book.cover_image,
            parent_id: book.parent_id,
            sort_order: book.sort_order,
            is_public: book.is_public,
            user_id: book.user_id_id,
            created_at: book.created_at.into(),
            updated_at: book.updated_at.into(),
        }
    }
}

/// 书籍树形结构响应
#[derive(Debug, Serialize, Deserialize)]
pub struct BookTreeResponse {
    pub book: BookResponse,
    pub children: Vec<BookTreeResponse>,
}

impl From<BookTree> for BookTreeResponse {
    fn from(tree: BookTree) -> Self {
        Self {
            book: BookResponse::from(tree.book),
            children: tree
                .children
                .into_iter()
                .map(BookTreeResponse::from)
                .collect(),
        }
    }
}

File: ui/src/services/books.ts
import { api } from '../lib/api'

export interface Book {
  id: number
  title: string
  description?: string
  cover_image?: string
  parent_id?: number
  sort_order?: number
  is_public?: boolean
  user_id: number
  created_at: string
  updated_at: string
  media_count?: number
  chapter_count?: number
}

export interface BookTree {
  book: Book
  children: BookTree[]
}

export interface CreateBookParams {
  title: string
  description?: string
  cover_image?: string
  parent_id?: number
  sort_order?: number
  is_public?: boolean
}

export interface UpdateBookParams {
  title?: string
  description?: string
  cover_image?: string
  parent_id?: number
  sort_order?: number
  is_public?: boolean
}

export interface ReorderBookParams {
  sort_order: number
}

export const booksService = {
  // 获取书籍列表
  async getBooks(): Promise<Book[]> {
    const response = await api.get('/books')
    return response.data
  },

  // 搜索书籍
  async searchBooks(query: string): Promise<Book[]> {
    const response = await api.get('/books/search', {
      params: { q: query }
    })
    return response.data
  },

  // 获取书籍详情
  async getBook(id: number): Promise<Book> {
    const response = await api.get(`/books/${id}`)
    return response.data
  },

  // 创建书籍
  async createBook(params: CreateBookParams): Promise<Book> {
    const response = await api.post('/books', params)
    return response.data
  },

  // 更新书籍
  async updateBook(id: number, params: UpdateBookParams): Promise<Book> {
    const response = await api.put(`/books/${id}`, params)
    return response.data
  },

  // 删除书籍
  async deleteBook(id: number): Promise<void> {
    await api.delete(`/books/${id}`)
  },

  // 获取书籍树形结构
  async getBookTree(id: number): Promise<BookTree> {
    const response = await api.get(`/books/${id}/tree`)
    return response.data
  },

  // 调整书籍顺序
  async reorderBook(id: number, params: ReorderBookParams): Promise<Book> {
    const response = await api.post(`/books/${id}/reorder`, params)
    return response.data
  }
}